File: 백준/Silver/16564. 히오스 프로게이머/히오스 프로게이머.py
import sys

n, k = map(int, sys.stdin.readline().split())
levels = []

for i in range(n):
    levels.append(int(sys.stdin.readline()))

levels.sort()

def can_up(levels, mid, k):
    sum = 0
    for i in range(n):
        if levels[i] < mid:
            sum += mid - levels[i]
    if sum <= k:
        return True
    return False


low = levels[0]
high = levels[-1] + k
result = 0
while low <= high:
    mid = (low + high) // 2
    if can_up(levels, mid, k):
        result = mid
        low = mid + 1
    else:
        high = mid - 1

print(result)

File: 백준/Silver/16564. 히오스 프로게이머/히오스 프로게이머.c
#include <stdio.h>
#include <stdlib.h>

int compare(const void* a, const void* b){
    return *(int*)a - *(int*)b;
}

int can_up(int* levels, int mid, int n, int k){
    int sum = 0;
    for (int i = 0; i < n; i++){
        if (levels[i] < mid) {
            sum += mid - levels[i];
            if (sum > k) {
                return 0;
            }
        }
    }
    return 1;
}

int main(void){
    int n, k;
    scanf("%d %d", &n, &k);

    int* levels = (int*)malloc(n * sizeof(int));
    for (int i = 0; i < n; i++){
        scanf("%d", &levels[i]);
    }

    qsort(levels, n, sizeof(int), compare);

    int low = levels[0];
    int high = levels[0] + k;
    int result = 0;
    while (low <= high){
        int mid = (low + high) / 2;
        if (can_up(levels, mid, n, k)){
            result = mid;
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    printf("%d\n", result);

    free(levels);

    return 0;
}